lib/src/rules/use_late_for_private_fields_and_variables.dart

```diff
@@ -0,0 +1,175 @@
+
+  CompilationUnitElement currentUnit;
+
+  static final lateables =
+      <CompilationUnitElement, List<VariableDeclaration>>{};
+  static final nullableAccess = <CompilationUnitElement, Set<Element>>{};
+
+  @override
+  void visitCompilationUnit(CompilationUnit node) {
+    if (node.featureSet.isEnabled(Feature.non_nullable)) {
+      currentUnit = node.declaredElement;
+      lateables.putIfAbsent(currentUnit, () => []);
+      nullableAccess.putIfAbsent(currentUnit, () => {});
+
+      super.visitCompilationUnit(node);
+
+      final units = node.declaredElement.library.units;
+      if (units.every(lateables.keys.contains)) {
+        _checkAccess(units);
+
+        // clean up
+        units.forEach((unit) {
+          lateables.remove(unit);
+          nullableAccess.remove(unit);
+        });
+      }
+    }
+  }
+
+  @override
+  void visitNode(AstNode node) {
+    var element = DartTypeUtilities.getCanonicalElementFromIdentifier(node);
+    if (element != null) {
+      var parent = node.parent;
+      if (parent is Expression) {
+        parent = (parent as Expression).unParenthesized;
+      }
+      if (node is SimpleIdentifier && node.inDeclarationContext()) {
+        // ok
+      } else if (parent is PostfixExpression &&
+          parent.operand == node &&
+          parent.operator.type == TokenType.BANG) {
+        // ok non-null access
+      } else if (parent is AssignmentExpression &&
+          parent.operator.type == TokenType.EQ &&
+          context.typeSystem.isNonNullable(parent.rightHandSide.staticType)) {
+        // ok non-null access
+      } else {
+        nullableAccess[currentUnit].add(element);
+      }
+    }
+    super.visitNode(node);
+  }
+
+  @override
+  void visitFieldDeclaration(FieldDeclaration node) {
+    for (var variable in node.fields.variables) {
+      final parent = node.parent;
+      if (Identifier.isPrivateName(variable.name.name) ||
+          _isPrivateNamedCompilationUnitMember(parent) ||
+          _isPrivateExtension(parent)) {
+        _visit(variable);
+      }
+    }
+  }
+
+  @override
+  void visitTopLevelVariableDeclaration(TopLevelVariableDeclaration node) {
+    for (var variable in node.variables.variables) {
+      if (Identifier.isPrivateName(variable.name.name)) {
+        _visit(variable);
+      }
+    }
+  }
+
+  void _visit(VariableDeclaration variable) {
+    if (variable.isLate) {
+      return;
+    }
+    if (variable.isSynthetic) {
+      return;
+    }
+    if (context.typeSystem.isNonNullable(variable.declaredElement.type)) {
+      return;
+    }
+    lateables[currentUnit].add(variable);
+  }
+
+  void _checkAccess(List<CompilationUnitElement> units) {
+    final allNullableAccess =
+        units.expand((unit) => nullableAccess[unit]).toSet();
+    for (final unit in units) {
+      for (var variable in lateables[unit]) {
+        if (!allNullableAccess.contains(variable.declaredElement)) {
+          rule.reportLint(variable);
```

Comment:
I believe that this will report the lint for variables declared in one file at the offset of the variable, but in the file currently being visited, which might be a different file. As far as I know, the linter doesn't have good support for reporting lints across a whole library.

What you'll probably need to do is use `reporter.reportError` and pass in an `AnalysisError` that is created here (rather than use `reportLint` which creates the `AnalysisError` for you).

We probably need some tests of libraries with parts to ensure that this is working as intended.